radixdlt-engine/src/main/java/com/radixdlt/atom/TxAction.java

```diff
@@ -64,5 +64,216 @@
 
 package com.radixdlt.atom;
 
-/** Marker interface for RadixEngine actions. */
-public interface TxAction {}
+import static com.radixdlt.utils.Strings.asEmptyIfNull;
+import static java.util.Objects.requireNonNull;
+
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.identifiers.REAddr;
+import com.radixdlt.utils.Bytes;
+import com.radixdlt.utils.UInt256;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.LongFunction;
+
+/** RadixEngine actions. */
+public sealed interface TxAction {
+  record BurnToken(REAddr resourceAddr, REAddr fromAddr, UInt256 amount) implements TxAction {
+    public BurnToken {
+      ensureNonZeroAmount(amount);
+    }
+  }
+
+  record CreateFixedToken(
+      REAddr resourceAddr,
+      REAddr accountAddr,
+      String symbol,
+      String name,
+      String description,
+      String iconUrl,
+      String tokenUrl,
+      UInt256 supply)
+      implements TxAction {
+    public CreateFixedToken {
+      ensureResourceAddressIsHashedKey(resourceAddr);
+      ensureNonZeroSupply(supply);
+      requireAllNonNull(accountAddr, symbol, name, description, iconUrl, tokenUrl);
+    }
+  }
+
+  record CreateMutableToken(
+      REAddr resourceAddress,
+      String symbol,
+      String name,
+      String description,
+      String iconUrl,
+      String tokenUrl,
+      ECPublicKey owner)
+      implements TxAction {
+    public CreateMutableToken(
+        REAddr resourceAddress,
+        String symbol,
+        String name,
+        String description,
+        String iconUrl,
+        String tokenUrl,
+        ECPublicKey owner) {
+      this.resourceAddress = resourceAddress;
+      this.symbol = symbol.toLowerCase();
+      this.owner = owner;
+
+      this.name = asEmptyIfNull(name);
+      this.description = asEmptyIfNull(description);
+      this.iconUrl = asEmptyIfNull(iconUrl);
+      this.tokenUrl = asEmptyIfNull(tokenUrl);
+    }
+  }
+
+  record CreateSystem(long timestamp) implements TxAction {}
+
+  record FeeReserveComplete(REAddr toAddr) implements TxAction {
+    public FeeReserveComplete {
+      ensureAddressIsAccount(toAddr);
+    }
+  }
+
+  record FeeReservePut(REAddr fromAddr, UInt256 amount) implements TxAction {}
+
+  record FeeReserveTake(REAddr toAddr, UInt256 amount) implements TxAction {}
+
+  record MintToken(REAddr resourceAddr, REAddr toAddr, UInt256 amount) implements TxAction {
+    public MintToken {
+      ensureNonZeroAmount(amount);
+    }
+  }
+
+  record NextEpoch(long timestamp) implements TxAction {}
+
+  record NextRound(
+      long view, boolean isTimeout, long timestamp, LongFunction<ECPublicKey> leaderMapping)
+      implements TxAction {}
```

Comment:
I think we're missing the nice ToString on this Action - although I'm not sure where it's used

Reply:
For each `record` compiler generates `toString()` method which is at least as good as one we had before. For example, here is output for `CreateSystem` action:
```
CreateSystem[timestamp=1]
```